### PR TITLE
New status queries for mip and balsamic

### PR DESCRIPTION
**Pull request by Mropat**
#### Modifies class StatusHandler:
Merge methods `cases_to_mip_analyze` and `cases_to_balsamic_analyze`, now called `cases_to_analyze` for any pipeline.  
Make method `cases_to_deliver` pipeline aware
Make method `cases_to_store` pipeline aware
Make method `cases_to_clean` pipeline aware

Add option to apply read count threshold for `cases_to_analyze` based on panel in ClinicalDB

Updated MIP and BALSAMIC workflows to utilize the new methods
Added tests for new methods in balsamic workflow

#### Removes class FastqAPI
FastqAPI was only used by BALSAMIC, and contained one method. The method is now moved to FastqHandler class. 

### How to prepare for test
- [x] ssh to hasta (depending on type of change)
- [x] install on stage:
`bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-cg-stage.sh [THIS-BRANCH-NAME]`
- [x] ssh to clinical-db (depending on type of change)
- [x] install on stage:
`bash /home/proj/production/servers/resources/clinical-db.scilifelab.se/update-clinical-api-stage.sh [THIS-BRANCH-NAME]`

### How to test
- [x] login to hasta
- [x] run `cg workflow balsamic start-avalable` and `cg workflow balsamic store-avalable`, verify the workflow starts, then cancel submitted jobs

### Expected test outcome
- [x] check that  analyses start and get stored automatically
- [x] Take a screenshot and attach or copy/paste the output.

No eligible analyses on stage right now :(

## Review
- [x] code approved by PG
- [x] tests executed by MR
- [x] "Merge and deploy" approved by PG
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a
- [x] **MINOR** - when you add functionality in a backwards compatible manner

### CODE REVIEW

**Comment by Mropat** [comment]
@patrikgrenfeldt  could you take a look at this?

File: tests/cli/workflow/balsamic/test_store_housekeeper.py [hunk 1]
```diff
@@ -126,3 +126,5 @@ def test_valid_case_already_added(
     assert result.exit_code != EXIT_SUCCESS
     # THEN user should be informed that bundle was already added
     assert "Bundle already added" in caplog.text
+    # THEN statusdb action should be set ot hold
+    assert balsamic_context["BalsamicAnalysisAPI"].store.family(case_id).action == "hold"
```
Comment (patrikgrenfeldt) [inline tests/cli/workflow/balsamic/test_store_housekeeper.py:130]: This action is by "tradition" used for holding the analysis and "Prodteam" will probably be confused if we use it in new ways

File: cg/cli/workflow/balsamic/base.py [hunk 2]
```diff
@@ -173,10 +172,14 @@ def store_housekeeper(context, case_id):
         balsamic_analysis_api.upload_bundle_housekeeper(case_id=case_id)
         LOG.info(f"Storing Analysis in ClinicalDB for {case_id}")
         balsamic_analysis_api.upload_analysis_statusdb(case_id=case_id)
-    except (BalsamicStartError, BundleAlreadyAddedError, FileExistsError) as e:
+    except (BundleAlreadyAddedError, FileExistsError) as e:
         LOG.error(f"Could not store bundle in Housekeeper and StatusDB: {e.message}!")
         balsamic_analysis_api.housekeeper_api.rollback()
         balsamic_analysis_api.store.rollback()
+        balsamic_analysis_api.set_statusdb_action(case_id=case_id, action="hold")
```
Comment (patrikgrenfeldt) [inline cg/cli/workflow/balsamic/base.py:179]: This action is by "tradition" used for holding the analysis and "Prodteam" will probably be confused if we use it in new ways, we should probably have a discussion on this type of change

**Review by patrikgrenfeldt — CHANGES_REQUESTED** [review]
Great work! The think I would like you to reconsider is the usage of the action attribute to represent a case as "failed to store" if I understand it correctly.

File: cg/apps/balsamic/fastq.py [hunk 3]
```diff
@@ -150,3 +150,58 @@ def _remove_files(files):
         for file in files:
             if os.path.isfile(file):
                 os.remove(file)
+
+    @staticmethod
+    def parse_header(line):
```
[the remark this answers is not in the document]
Comment (Mropat) [reply cg/apps/balsamic/fastq.py:155]: Yes, in the next PR! This warrants a global refactor of fastq api for both mip and balsamic (perhaps other workflows). Changing this here will cause merge conflicts with my other PR also. I believe its better to do it as its own undertaking

File: tests/cli/workflow/balsamic/conftest.py [hunk 4]
```diff
@@ -359,13 +365,15 @@ def balsamic_store(base_store: Store, helpers) -> Store:
         is_tumour=True,
         application_type="wgs",
         data_analysis="balsamic",
+        reads=10,
```
[the remark this answers is not in the document]
Comment (patrikgrenfeldt) [reply tests/cli/workflow/balsamic/conftest.py:368]: This became weird as a 'suggestion', please let me know if I need to clarify :)

File: cg/cli/workflow/balsamic/base.py (hunk 2 above)
Comment (Mropat) [reply cg/cli/workflow/balsamic/base.py:179]: Will remove for now until a suitable solution introduced in ClinicalDB

File: cg/meta/workflow/balsamic.py [hunk 5]
```diff
@@ -381,8 +386,52 @@ def upload_analysis_statusdb(self, case_id: str) -> None:
         self.store.add_commit(new_analysis)
         LOG.info(f"Analysis successfully stored in ClinicalDB: {case_id} : {analysis_start}")
 
+    def family_has_correct_samples(self, case_id: str) -> bool:
```
[the remark this answers is not in the document]
Comment (Mropat) [reply cg/meta/workflow/balsamic.py:389]: :+1:

File: cg/store/api/status.py [hunk 6]
```diff
@@ -59,18 +59,15 @@ def samples_to_sequence(self):
         )
         return records
 
-    def cases_to_mip_analyze(self, limit: int = 50):
-        """Fetch families without analyses where all samples are sequenced."""
-
-        families_q = (
+    def cases_to_analyze(
+        self, pipeline: str = "", threshold: bool = False, limit: int = 100000
```
[the remark this answers is not in the document]
Comment (Mropat) [reply cg/store/api/status.py:63]: :+1:

File: cg/store/api/status.py [hunk 7]
```diff
@@ -533,7 +520,25 @@ def _all_samples_have_sequence_data(links: List[models.FamilySample]) -> bool:
...
+                > self.Application.query.filter_by(
+                    id=self.ApplicationVersion.query.filter_by(
+                        id=link.sample.application_version_id
+                    )
+                    .first()
+                    .application_id
+                )
+                .first()
+                .target_reads
+                * 0.75
```
[the remark this answers is not in the document]
Comment (Mropat) [reply cg/store/api/status.py:536]: :+1:

File: tests/cli/workflow/balsamic/conftest.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Mropat) [reply tests/cli/workflow/balsamic/conftest.py:368]: Created a new case for this with the desired naming convention

**Comment by Mropat** [comment]
@patrikgrenfeldt I addressed the suggestions, please take another look!

**Comment by Mropat** [comment]
Automatic start 
image

**Comment by Mropat** [comment]
storing success
image
image
image

**Comment by Mropat** [comment]
starting multiple
image
Here it finds case without panel in LIMS and still continues
image

**Comment by patrikgrenfeldt** [comment]
Deployed on Hasta:
image

Deployed on clinical-db:
image

**Pull request merged**